src/server/dispatch.rs

```diff
@@ -0,0 +1,136 @@
+            )*
+        }
+
+        $(
+            impl From<(Request<$request_type>, InitActionContext)> for DispatchRequest {
+                fn from((req, ctx): (Request<$request_type>, InitActionContext)) -> Self {
+                    DispatchRequest::$request_type($request_type::new(), req, ctx)
+                }
+            }
+        )*
+
+        impl DispatchRequest {
+            fn handle<O: Output>(self, out: &O) {
+                match self {
+                $(
+                    DispatchRequest::$request_type(mut var, req, ctx) => {
+                        let Request { id, params, received, .. } = req;
+                        let fallback = var.fallback_response();
+                        let timeout = var.timeout();
+
+                        let receiver = receive_from_thread(move || {
+                            if received.elapsed() >= timeout {
+                                var.fallback_response()
+                            }
+                            else {
+                                var.handle(ctx, params)
+                            }
+                        });
+
+                        match receiver.recv_timeout(timeout).unwrap_or(fallback) {
+                            Ok(response) => response.send(id, out),
+                            Err(ResponseError::Empty) => debug!("Error handling request"),
+                            Err(ResponseError::Message(code, msg)) => {
+                                out.failure_message(id, code, msg)
+                            }
+                        }
+                    }
+                )*
+                }
+            }
+        }
+    }
+}
+
+define_dispatch_request_enum!(
+    Completion,
+    Definition,
+    References,
+    WorkspaceSymbol,
+    Symbols,
+    Hover,
+    FindImpls,
+    DocumentHighlight,
+    Rename,
+    CodeAction
+);
+
+/// Provides ability to dispatch requests to a worker thread that will
+/// handle the requests sequentially, without blocking stdin.
+/// Requests dispatched this way are automatically timed out & avoid
+/// processing if have already timed out before starting.
+pub struct Dispatcher {
+    sender: mpsc::Sender<DispatchRequest>,
+}
+
+impl Dispatcher {
+    pub fn new<O: Output>(out: O) -> Self {
+        let (sender, receiver) = mpsc::channel::<DispatchRequest>();
+
+        thread::Builder::new().name("dispatch-worker".into()).spawn(move || {
+            while let Ok(request) = receiver.recv() {
+                request.handle(&out)
+            }
+        }).unwrap();
+
+        Self { sender }
+    }
+
+    pub fn dispatch<R: Into<DispatchRequest>>(&self, request: R) {
+        if let Err(err) = self.sender.send(request.into()) {
+            debug!("Failed to dispatch request: {:?}", err);
+        }
+    }
+}
+
+/// Stdin-nonblocking request logic designed to be packed into a `DispatchRequest`
+/// and handled on the `WORK_POOL` via a `Dispatcher`.
```

Comment:
AIUI, we're not actually using a WORK_POOL here?

Reply:
In the handle method `receive_from_thread` receives results from the `WORK_POOL` and that's because we want to be able to timeout.